cleo_plugins/FileSystemOperations/FileSystemOperations.cpp

```diff
@@ -58,6 +85,94 @@ class FileSystemOperations
         return path;
     }
 
+    //0A9A=3,%3d% = openfile %1d% mode %2d% // IF and SET
+    static OpcodeResult WINAPI opcode_0A9A(CRunningScript* thread)
+    {
+        auto filename = ReadPathParam(thread);
+
+        char mode[16];
+        auto paramType = CLEO_GetOperandType(thread);
+        if (IsImmInteger(paramType) || IsVariable(paramType))
+        {
+            // integer param (for backward compatibility with CLEO 3)
+            union
+            {
+                DWORD uParam;
+                char strParam[4];
+            } param;
+            param.uParam = CLEO_GetIntOpcodeParam(thread);
+            strcpy(mode, param.strParam);
+        }
+        else
+        {
+            CLEO_ReadStringOpcodeParam(thread, mode, sizeof(mode));
+        }
+
+        // either CLEO 3 or CLEO 4 made a big mistake! (they differ in one major unapparent preference)
+        // lets try to resolve this with a legacy mode
+        bool legacy = CLEO_GetScriptVersion(thread) < CLEO_VER_4_3;
+
+        auto handle = File::open(filename.c_str(), mode, legacy);
+        if (!File::isOk(handle))
+        {
+            CLEO_SetIntOpcodeParam(thread, NULL);
+            CLEO_SetThreadCondResult(thread, false);
+            return OR_CONTINUE;
+        }
+
+        m_hFiles.insert(handle);
+        CLEO_SetIntOpcodeParam(thread, handle);
+        CLEO_SetThreadCondResult(thread, true);
+        return OR_CONTINUE;
+    }
+
+    //0A9B=1,closefile %1d%
+    static OpcodeResult WINAPI opcode_0A9B(CRunningScript* thread)
+    {
+        DWORD handle = CLEO_GetIntOpcodeParam(thread); OPCODE_VALIDATE_HANDLE(handle)
```

Comment:
this probably can be extracted to another macro or method: 

DWORD handle = GetFileHandle(thread); (calls CLEO_GetIntOpcodeParam & does OPCODE_VALIDATE_HANDLE)

Reply:
Method no, as it uses macro so can call "return" of the nesting fiction.
Maybe it is not good idea to over engineering it. Macros are dirty and do not play nicely when debugging.

Reply:
yea I mean you use the same pair of commands 12 times, could extract into a small function to reduce code duplicating